confmap/expand.go

```diff
@@ -0,0 +1,197 @@
+	switch v := value.(type) {
+	case string:
+		if !strings.Contains(v, "${") || !strings.Contains(v, "}") {
+			// No URIs to expand.
+			return value, false, nil
+		}
+
+		if strings.Count(v, "}") > 100 {
+			// Too many closing brackets. Don't expand to protect from too deep recursion.
+			return value, false, nil
+		}
+
+		uri := findURI(v)
+		if uri != "" && uri == value {
+			// If the value is a single URI, then the return value can be anything.
+			// This is the case `foo: ${file:some_extra_config.yml}`.
+			return mr.expandStringURI(ctx, v)
+		}
+
+		// Embedded or nested URIs.
+		return mr.findAndExpandURI(ctx, v)
+	case []any:
+		nslice := make([]any, 0, len(v))
+		nchanged := false
+		for _, vint := range v {
+			val, changed, err := mr.expandValue(ctx, vint)
+			if err != nil {
+				return nil, false, err
+			}
+			nslice = append(nslice, val)
+			nchanged = nchanged || changed
+		}
+		return nslice, nchanged, nil
+	case map[string]any:
+		nmap := map[string]any{}
+		nchanged := false
+		for mk, mv := range v {
+			val, changed, err := mr.expandValue(ctx, mv)
+			if err != nil {
+				return nil, false, err
+			}
+			nmap[mk] = val
+			nchanged = nchanged || changed
+		}
+		return nmap, nchanged, nil
+	}
+	return value, false, nil
+}
+
+// findURI finds the URI corresponding to the first closing bracket in input. It returns
+// the URI if it is expandable, or an empty string if it is not expandable.
+// Note: findURI is only called when input contains a closing bracket.
+func findURI(input string) string {
+	closeIndex := strings.Index(input, "}")
+	openIndex := strings.LastIndex(input[:closeIndex+1], "${")
+	if openIndex < 0 {
+		// Should not expand because there is a missing ${.
+		return ""
+	}
+
+	uri := input[openIndex : closeIndex+1]
+
+	if !strings.Contains(uri, ":") {
+		// Should not expand. This is expanded in the expandconverter.
+		return ""
+	}
+
+	return uri
+}
+
+// findAndExpandURI attempts to find and expand the first occurrence of an expandable URI in input.
+func (mr *Resolver) findAndExpandURI(ctx context.Context, input string) (output string, changed bool, err error) {
+	var repl string
+	uri := findURI(input)
+	if uri != "" {
+		repl, changed, err = mr.expandURI(ctx, uri)
+		input = strings.ReplaceAll(input, uri, repl)
+	} else {
+		// Check if the next URI in input is expandable.
+		input, changed, err = mr.nextExpandableURI(ctx, input)
+	}
+	return input, changed, err
+}
+
+// nextExpandableURI attempts to find and expand the next expandable URI in input. nextExpandableURI is called
+// when the first occurrence of URI in input is not expandable.
+func (mr *Resolver) nextExpandableURI(ctx context.Context, input string) (string, bool, error) {
+	var err error
+	var expandedRemaining string
+	var changed bool
+
+	closeIndex := strings.Index(input, "}")
+	noExpand := input[:closeIndex+1]
+	remaining := input[closeIndex+1:]
+
+	// if remaining does not contain }, there are no URIs left: stop recursion.
```

Comment:
This is not what the code does :) it is opposite, but I would like the code to be like this:

 ```golang
	if !strings.Contains(remaining, "}") {
		return input, changed, err
	}
	expandedRemaining, changed, err = mr.findAndExpandURI(ctx, remaining)
	return noExpand + expandedRemaining, changed, err
```

Reply:
Yeah, good catch :) Made the code as suggested 👍